File: C/array/stack.c
/**
 * Author: Axel Ariel Saravia
 */

/*
    IMPORTANT

    the T type is the signed type

    stack struct is {
        T* content;
        size_t capacity;
        size_t length;
    }
*/

# include "stack.h"

stack stack_create() {
    T* content = calloc(8, sizeof(T));

    return (
        !content
        ? (stack){0}
        : (stack){
            .content = content,
            .capacity = 8,
            .length = 0,
        }
    );
}

T stack_get(stack const s, size_t const i) {
    T res = 0; 
    if (i < s.length) {
        res = s.content[i];
    }
    return res;
}

T stack_set(stack ps[static 1] , size_t const i, T const x) {
    T y = 0;
    if (i < ps->length) {
        y = ps->content[i];
        ps->content[i] = x;
    }
    return y;
}

stack stack_add(stack s, size_t i, T const x) {
    if (s.length < i) {
        i = s.length;
    }
    if (s.length == s.capacity) {
        s = stack_resize(s);
    }
    for (size_t j = s.length; i < j; j += 1) {
        s.content[j] = s.content[j - 1];
    }
    s.content[i] = x;
    s.length += 1;
    return s;
}

stack stack_remove(stack s, size_t i) {
    size_t const last = s.length - 1;
    if (s.length <= i) {
        i = s.length - 1;
    }
    while (i < last) {
        s.content[i] = s.content[i + 1];
        i += 1;
    }
    s.length -= 1;
    if (8 < s.capacity && 3 * s.length <= s.capacity) {
        s = stack_resize(s);
    }
    return s;
}

stack stack_resize(stack s) {
    size_t j = 2 * s.length;
    size_t new_capacity = (8 < j ? j : 8);
    s.content = realloc(s.content, sizeof(T[new_capacity]));
    s.capacity = new_capacity;
    return s;
}

void stackRef_add(stack ps[static 1], size_t const i, T const x) {
    *ps = stack_add(*ps, i, x);
}

void stackRef_remove(stack ps[static 1], size_t i) {
    *ps = stack_remove(*ps, i);
}

void stack_free(stack ps[static 1]) {
    free(ps->content);
    *ps = (stack){0};
}

File: Go/arrayStack/stack.go
/**
* Author: Axel Ariel Saravia
*/

package arrayStack

type Stack[T any] struct {
    Len uint
    content []T
}

func CreateStack[T any]() Stack[T] {
    return Stack[T]{
        Len: 0,
        content: make([]T, 1),
    }
}

func (a *Stack[T]) Get(i uint) T {
    return a.content[i]
}

func (a *Stack[T]) Set(i uint, x T) T {
    v := a.content[i]
    a.content[i] = x
    return v
}

func (a *Stack[T]) Add(i uint, x T) *Stack[T] {
    if (a.Len < i) {
        i = a.Len
    }
    if (a.Len == uint(cap(a.content))) {
        resize(a)
    }
    for j := a.Len; i < j; j -= 1 {
        a.content[j] = a.content[j - 1]
    }
    a.content[i] = x
    a.Len += 1
    return a
}

func (a *Stack[T]) Remove(i uint) *Stack[T] {
    last := a.Len - 1
    if (a.Len <= i) {
        i = last
    }
    for i < last {
        a.content[i] = a.content[i + 1] 
        i += 1
    }
    a.Len -= 1
    if (3 * a.Len <= uint(cap(a.content))) {
        resize(a)
    }
    return a
}

func (a *Stack[T]) Clear() {
    a.Len = 0
    a.content = a.content[:1]
}

func resize[T any](a *Stack[T]) {
    var newCapacity uint
    if c:= 2 * a.Len; 1 < c {
        newCapacity = c
    } else {
        newCapacity = 1
    }
    n := make([]T, newCapacity)
    copy(n, a.content)
    a.content = n
}
